File: frontend/src/precompute.py
import csv
import json
import re
from collections import defaultdict
from geopy.distance import geodesic

# Pattern for extracting coordinates from POINT and LINESTRING entries
coord_pattern = re.compile(r"\((.+)\)")

# Load the stops data
stops = {}
with open('cairo_metro_stops.csv', 'r') as f:
    reader = csv.DictReader(f)
    for row in reader:
        coordinates = tuple(map(float, coord_pattern.search(row['geometry']).group(1).split()))
        stops[row['stop_name']] = {'coordinates': coordinates, 'connections': defaultdict(list)}

# Function to calculate distance between two coordinates
def calculate_distance(coord1, coord2):
    return geodesic(coord1, coord2).kilometers

# Load the routes data
with open('cairo_metro_trips.csv', 'r') as f:
    reader = csv.DictReader(f)
    for row in reader:
        route_coordinates = [tuple(map(float, coords.split())) for coords in coord_pattern.findall(row['geometry'])[0].split(",")]
        for i in range(len(route_coordinates) - 1):
            start_coords, end_coords = route_coordinates[i], route_coordinates[i+1]
            start_station = min(stops, key=lambda s: calculate_distance(stops[s]['coordinates'], start_coords))
            end_station = min(stops, key=lambda s: calculate_distance(stops[s]['coordinates'], end_coords))

            # Update the stations' connections
            if start_station != end_station:  # Ensure the station is not connected to itself
                if end_station not in stops[start_station]['connections']:
                    stops[start_station]['connections'][end_station] = {'route_id': row['route_id'], 'distance': calculate_distance(stops[start_station]['coordinates'], stops[end_station]['coordinates'])}
                if start_station not in stops[end_station]['connections']:
                    stops[end_station]['connections'][start_station] = {'route_id': row['route_id'], 'distance': calculate_distance(stops[start_station]['coordinates'], stops[end_station]['coordinates'])}

# Output the graph to a JSON file
with open('metro_graph.json', 'w') as f:
    json.dump(stops, f)

File: backend/preCompute.js
const mongoose = require('mongoose');
const geolib = require('geolib');

const mongoURL = "mongodb+srv://alhassan:p7OcHSJ3LRpJTgHz@stations.r5pneer.mongodb.net/stations?retryWrites=true&w=majority";

// Define schemas and models
const stationSchema = new mongoose.Schema({
    FID: {
        type: String,
        required: true,
    },
    fid: {
        type: Number,
        required: true,
    },
    geometry: {
        type: String,
        required: true,
    },
    stop_id: {
        type: String,
        required: true,
    },
    stop_name: {
        type: String,
        required: true,
    },
});

const routeSchema = new mongoose.Schema({
    FID: {
        type: String,
        required: true,
    },
    fid: {
        type: Number,
        required: true,
    },
    geometry: {
        type: String,
        required: true,
    },
    route_id: {
        type: String,
        required: true,
    },
    route_short_name: {
        type: String,
        required: true,
    },
    route_long_name: {
        type: String,
        required: true,
    },
    route_headsign: {
        type: String,
        required: true,
    },
    trip_id: {
        type: String,
        required: true,
    },
    destination: {
        type: String,
        required: true,
    },
});

const Station = mongoose.model("Station", stationSchema);
const Route = mongoose.model("Route", routeSchema);

// Function to calculate distance between two coordinates
const calculateDistance = (coord1, coord2) => {
    return geolib.getDistance(
        { latitude: coord1[0], longitude: coord1[1] },
        { latitude: coord2[0], longitude: coord2[1] }
    ) / 1000; // Conversion from meters to kilometers
};

mongoose
  .connect(mongoURL, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
  })
  .then(async () => {
    console.log("database connected.");
    
    let stops = {};

    const stations = await Station.find({});
    for (let station of stations) {
        let coordinates = station.geometry.match(coordPattern)[1].split(' ').map(parseFloat);
        stops[station.stop_name] = { 'coordinates': coordinates, 'connections': {} };
    }

    const routes = await Route.find({});
    for (let route of routes) {
        let routeCoordinates = route.geometry.match(coordPattern)[0].split(',').map(coords => coords.split(' ').map(parseFloat));
        for (let i = 0; i < routeCoordinates.length - 1; i++) {
            let startCoords = routeCoordinates[i];
            let endCoords = routeCoordinates[i+1];
            let startStation = Object.keys(stops).reduce((a, b) => calculateDistance(stops[a]['coordinates'], startCoords) < calculateDistance(stops[b]['coordinates'], startCoords) ? a : b);
            let endStation = Object.keys(stops).reduce((a, b) => calculateDistance(stops[a]['coordinates'], endCoords) < calculateDistance(stops[b]['coordinates'], endCoords) ? a : b);

            if (startStation != endStation) {
                if (!stops[startStation]['connections'].hasOwnProperty(endStation))
                    stops[startStation]['connections'][endStation] = { 'route_id': route.route_id, 'distance': calculateDistance(stops[startStation]['coordinates'], stops[endStation]['coordinates']) };
                if (!stops[endStation]['connections'].hasOwnProperty(startStation))
                    stops[endStation]['connections'][startStation] = { 'route_id': route.route_id, 'distance': calculateDistance(stops[startStation]['coordinates'], stops[endStation]['coordinates']) };
            }
        }
    }

    fs.writeFileSync('metro_graph.json', JSON.stringify(stops, null, 4));
  });
